src/datatype/functions_svecf32.rs

```diff
@@ -1,7 +1,504 @@
+    count: u64,
+    phantom: [F32; 0],
+}
+
+impl SparseAccumulateStateHeader {
+    fn varlena(size: usize) -> u32 {
+        (size << 2) as u32
+    }
+    fn layout(capacity: usize) -> Layout {
+        u32::try_from(capacity).expect("Vector is too large.");
+        let layout_alpha = Layout::new::<SparseAccumulateStateHeader>();
+        let layout1 = Layout::array::<u32>(capacity).unwrap();
+        let layout2 = Layout::array::<F32>(capacity).unwrap();
+        let layout = layout_alpha.extend(layout1).unwrap().0.pad_to_align();
+        layout.extend(layout2).unwrap().0.pad_to_align()
+    }
+    pub fn dims(&self) -> usize {
+        self.dims as usize
+    }
+    pub fn len(&self) -> usize {
+        self.len as usize
+    }
+    pub fn capacity(&self) -> usize {
+        self.capacity as usize
+    }
+    pub fn count(&self) -> u64 {
+        self.count
+    }
+    /// Get the indexes of the sparse state.
+    fn indexes(&self) -> &[u32] {
+        let ptr = self.phantom.as_ptr().cast();
+        unsafe { std::slice::from_raw_parts(ptr, self.len()) }
+    }
+    /// Get the values of the sparse state.
+    fn values(&self) -> &[F32] {
+        unsafe {
+            let ptr = self.phantom.as_ptr().cast::<u32>().add(self.capacity());
+            let offset = ptr.align_offset(8);
+            let ptr = ptr.add(offset).cast();
+            std::slice::from_raw_parts(ptr, self.len())
+        }
+    }
+    /// Get the mutable references of the indexes and values of the sparse state. The indexes and values may contain reserved elements.
+    fn indexes_values_mut(&mut self) -> (&mut [u32], &mut [F32]) {
+        let ptr = self.phantom.as_mut_ptr().cast();
+        let indexes = unsafe { std::slice::from_raw_parts_mut(ptr, self.capacity as usize) };
+        let len = self.capacity as usize;
+        let ptr = unsafe { self.phantom.as_mut_ptr().cast::<u32>().add(len) };
+        let offset = ptr.align_offset(8);
+        let ptr = unsafe { ptr.add(offset).cast() };
+        let values = unsafe { std::slice::from_raw_parts_mut(ptr, len) };
+        (indexes, values)
+    }
+}
+
+pub struct SparseAccumulateStateBorrowed<'a> {
+    #[allow(dead_code)]
+    dims: u32,
+    len: u32,
+    capacity: u32,
+    #[allow(dead_code)]
+    count: u64,
+    indexes: &'a mut [u32],
+    values: &'a mut [F32],
+}
+
+impl<'a> SparseAccumulateStateBorrowed<'a> {
+    pub fn new(
+        dims: u32,
+        len: u32,
+        capacity: u32,
+        count: u64,
+        indexes: &'a mut [u32],
+        values: &'a mut [F32],
+    ) -> Self {
+        Self {
+            dims,
+            len,
+            capacity,
+            count,
+            indexes,
+            values,
+        }
+    }
+
+    /// Try to merge a sparse vector into the state in place.
+    /// Return the length of the merged state if success, otherwise return an error.
+    pub fn merge_in_place(&mut self, svec: SVecf32Borrowed<'_>) -> Result<u32, AccumulateError> {
```

Comment:
We could grow the state util there are more empty slows than values in `svec` so that `merge_in_place` never fails. It's simpler.

Reply:
Right, it makes the code more readable.